File: apps/go-api/internal/handler/insights.go
package handler

import (
	"net/http"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"
	mw "github.com/reche/gravity-room/apps/go-api/internal/middleware"
	"github.com/reche/gravity-room/apps/go-api/internal/model"
	"github.com/reche/gravity-room/apps/go-api/internal/service"
)

// InsightsHandler serves pre-computed analytics insights.
type InsightsHandler struct {
	Pool *pgxpool.Pool
}

// HandleList handles GET /api/insights?types=volume_trend,frequency,...
// Returns the user's pre-computed insights, optionally filtered by type.
func (h *InsightsHandler) HandleList(w http.ResponseWriter, r *http.Request) {
	userID := mw.UserID(r.Context())

	var types []string
	if raw := r.URL.Query().Get("types"); raw != "" {
		for _, t := range strings.Split(raw, ",") {
			if t = strings.TrimSpace(t); t != "" {
				types = append(types, t)
			}
		}
	}

	insights, err := service.GetInsights(r.Context(), h.Pool, userID, types)
	if err != nil {
		respondJSON(w, r, http.StatusInternalServerError, model.ErrorResponse{
			Error: "failed to fetch insights",
			Code:  "internal_error",
		})
		return
	}

	respondJSON(w, r, 0, model.InsightsListResponse{Data: insights})
}

File: apps/api/src/routes/insights.ts
import { Elysia, t } from 'elysia';
import { jwtPlugin, resolveUserId } from '../middleware/auth-guard';
import { rateLimit } from '../middleware/rate-limit';
import { requestLogger } from '../middleware/request-logger';
import { getInsights } from '../services/insights';
import { ApiError } from '../middleware/error-handler';
import { INSIGHT_TYPES, parseInsightTypesQuery } from '../lib/insight-types';
import { logger } from '../lib/logger';

const security = [{ bearerAuth: [] }];

export const insightsRoutes = new Elysia({ prefix: '/insights' })
  .use(requestLogger)
  .use(jwtPlugin)
  .resolve(resolveUserId)

  .get(
    '/',
    async ({ userId, query }) => {
      const parsed = parseInsightTypesQuery(query.types);
      if (!parsed.ok) {
        throw new ApiError(400, 'Invalid insight type', 'INVALID_INSIGHT_TYPE', {
          details: {
            invalidValues: parsed.error.invalidValues,
            validValues: INSIGHT_TYPES,
          },
        });
      }

      await rateLimit(userId, 'GET /insights', { maxRequests: 30 });

      try {
        const rows = await getInsights(userId, parsed.value);
        return {
          data: rows.map((r) => ({
            insightType: r.insightType,
            exerciseId: r.exerciseId,
            payload: r.payload,
            computedAt: r.computedAt.toISOString(),
            validUntil: r.validUntil ? r.validUntil.toISOString() : null,
          })),
        };
      } catch (err) {
        logger.error({ err, userId }, 'failed to fetch insights');
        throw err;
      }
    },
    {
      query: t.Object({
        types: t.Optional(t.String()),
      }),
      detail: {
        tags: ['Insights'],
        summary: 'List user insights',
        description:
          'Returns pre-computed analytics insights for the authenticated user. ' +
          'Optionally filter by comma-separated insight types. ' +
          'Valid values: volume_trend, frequency, plateau_detection, load_recommendation. ' +
          'Unknown types return 400 with { code: "INVALID_INSIGHT_TYPE", invalidValues, validValues }.',
        security,
      },
    }
  );
